File: day-three.py
with open('day-three.txt') as f:
    lines = f.read().splitlines()

line_length = len(lines[0])
joltage_list = []

for line in lines:
    curr_max_num = 0
    curr_max_lead_digit = 0

    for i in range(line_length - 1):
        curr_lead_digit = int(line[i])

        if curr_lead_digit > curr_max_lead_digit:
            curr_max_lead_digit = curr_lead_digit

            for j in range(i + 1, line_length):
                curr_digit = int(line[j])

                curr_num = curr_lead_digit * 10 + curr_digit

                if curr_num > curr_max_num:
                    curr_max_num = curr_num

    joltage_list.append(curr_max_num)

print(sum(joltage_list))


### Claude improved version

def max_k_digit_number(line, k):
    """Find the max k-digit number from any k digits in order."""
    n = len(line)
    digits = []
    cursor = 0

    for i in range(k):
        # Valid range: [cursor, n - (k - 1 - i))
        # Need to leave room for (k - 1 - i) more digits after this pick
        end = n - (k - 1 - i)

        # Find leftmost maximum in line[cursor:end]
        best_pos = cursor
        for j in range(cursor + 1, end):
            if line[j] > line[best_pos]:
                best_pos = j

        digits.append(line[best_pos])
        cursor = best_pos + 1

    return int(''.join(digits))

print(sum(max_k_digit_number(line, 2) for line in lines))   # Part 1
print(sum(max_k_digit_number(line, 12) for line in lines))  # Part 2

File: day-three.exs
defmodule DayThree do
  @moduledoc """
  Find the maximum k-digit number by selecting k digits in order from each line.
  """

  def parse_lines(filename) do
    filename
    |> File.read!()
    |> String.trim()
    |> String.split("\n")
  end

  def max_k_digit_number(line, k) do
    # Convert to tuple for O(1) indexing (String.at is O(n)!)
    chars = line |> String.to_charlist() |> List.to_tuple()
    n = tuple_size(chars)

    chars
    |> pick_digits(n, k, 0)
    |> List.to_string()
    |> String.to_integer()
  end

  defp pick_digits(_chars, _n, 0, _cursor), do: []

  defp pick_digits(chars, n, remaining, cursor) do
    # Valid range: [cursor, n - remaining + 1)
    end_pos = n - remaining + 1
    best_pos = find_leftmost_max(chars, cursor, end_pos)

    [elem(chars, best_pos) | pick_digits(chars, n, remaining - 1, best_pos + 1)]
  end

  defp find_leftmost_max(chars, start, end_pos) do
    Enum.reduce((start + 1)..(end_pos - 1)//1, start, fn i, best ->
      if elem(chars, i) > elem(chars, best), do: i, else: best
    end)
  end

  def main do
    lines = parse_lines("day-three.txt")

    part_one = lines |> Enum.map(&max_k_digit_number(&1, 2)) |> Enum.sum()
    part_two = lines |> Enum.map(&max_k_digit_number(&1, 12)) |> Enum.sum()

    IO.puts(part_one)
    IO.puts(part_two)
  end
end

DayThree.main()
